src/plugin.rs

```diff
@@ -50,12 +53,102 @@ impl Serialize for Error {
 
 type Result<T> = std::result::Result<T, Error>;
 
+/// Separates a "connection string" into it's constituant parts
+/// which are: (a) the database, and (b) the connection details.
+///
+/// In most cases you connect with network details but for a SQLite
+/// database it is with a file based resource where the file based resource
+/// will need to be combined with the app's designated directory in filesystem.
+enum Connection {
+    Sqlite(String, Option<PathBuf>),
+    Mysql(String),
+    Postgres(String),
+}
+
+impl Connection {
+    fn add_app_directory(self, app_dir: &PathBuf) -> Self {
+        match self {
+            Connection::Sqlite(d, _) => Connection::Sqlite(d.clone(), Some(app_dir.clone())),
+            _ => self,
+        }
+    }
+}
+
+impl From<&str> for Connection {
+    fn from(connect: &str) -> Self {
+        let (db_type, details) = connect
+            .split_once(":")
+            .expect("Couldn't parse the connection string for the database! Please ensure that your connection string has a ':' character in it.");
+        let db_type = String::from(db_type).to_lowercase();
+
+        match db_type.as_str() {
+            "sqlite" => Connection::Sqlite(details.to_string(), None),
+            "mysql" => Connection::Mysql(details.to_string()),
+            "postgres" => Connection::Postgres(details.to_string()),
+            _ => panic!("The connection string passed in has an unrecognized database and can not be used: {}", details)
+        }
+    }
+}
+
+impl From<String> for Connection {
+    fn from(connect: String) -> Self {
+        let s = connect.as_str();
+        Connection::from(s)
+    }
+}
+
+impl ToString for Connection {
+    fn to_string(&self) -> String {
+        match &self {
+            Connection::Sqlite(file, Some(app_dir)) => String::from(
+                ["sqlite:", app_dir.join(&file).to_str().unwrap_or(&file)]
+                    .concat()
+                    .to_string(),
+            ),
+            Connection::Sqlite(file, None) => String::from(["sqlite:", &file].concat().to_string()),
+            Connection::Mysql(file) => String::from(["mysql:", &file].concat().to_string()),
+            Connection::Postgres(file) => String::from(["postgres:", &file].concat().to_string()),
+        }
+    }
+}
+
 #[derive(Default)]
 struct DbInstances(Mutex<HashMap<String, Pool<Db>>>);
 
+/// Provides important contextual path locations for the runtime environment.
+struct Locations {
```

Comment:
The `Locations` struct provides a place to put the BufPath's for app and resource directories as well as I have added some traits to make it more ergonomic.

Reply:
The locations struct is no longer. Long live the locations struct.